unixfs/src/file/adder.rs

```diff
@@ -0,0 +1,432 @@
+use quick_protobuf::{MessageWrite, Writer};
+use std::borrow::Cow;
+use std::fmt;
+use std::num::NonZeroUsize;
+
+use sha2::{Digest, Sha256};
+
+/// File tree builder
+pub struct FileAdder {
+    chunker: Chunker,
+    block_buffer: Vec<u8>,
+    // the index in the outer vec is the height (0 == leaf, 1 == first links to leafs, 2 == links
+    // to first link blocks)
+    unflushed_links: Vec<Vec<(Cid, u64, u64)>>,
+}
+
+impl std::default::Default for FileAdder {
+    fn default() -> Self {
+        Self::with_chunker(Chunker::Size(1024 * 256))
+    }
+}
+
+impl fmt::Debug for FileAdder {
+    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(
+            fmt,
+            "FileAdder {{ chunker: {:?}, block_buffer: {}/{}, unflushed_links: {} }}",
+            self.chunker,
+            self.block_buffer.len(),
+            self.block_buffer.capacity(),
+            self.unflushed_links
+                .iter()
+                .map(|l| l.len().to_string())
+                .collect::<Vec<_>>()
+                .join("/")
+        )
+    }
+}
+
+impl FileAdder {
+    /// Creates `FileAdder` with the given chunker. Typically one could just call
+    /// `FileAdder::default()`.
+    pub fn with_chunker(chunker: Chunker) -> Self {
+        let hint = chunker.size_hint();
+        FileAdder {
+            chunker,
+            block_buffer: Vec::with_capacity(hint),
+            unflushed_links: Default::default(),
+        }
+    }
+
+    /// Returns a likely amount of buffering the file adding works the best.
+    pub fn size_hint(&self) -> usize {
+        self.chunker.size_hint()
+    }
+
+    /// Called to push new file bytes into the tree builder.
+    ///
+    /// Returns the newly created blocks (at most 2) and their respective Cids, and the amount of
+    /// `input` consumed.
+    pub fn push(
+        &mut self,
+        input: &[u8],
+    ) -> Result<(impl Iterator<Item = (Cid, Vec<u8>)>, usize), ()> {
+        // case 0: full chunk is not ready => empty iterator, full read
+        // case 1: full chunk becomes ready, maybe short read => at least one block
+        //     1a: not enough links => iterator of one
+        //     1b: link block is ready => iterator of two blocks
+
+        let (accepted, ready) = self.chunker.accept(input, &self.block_buffer);
+        self.block_buffer.extend_from_slice(accepted);
+        let written = accepted.len();
+
+        let (leaf, links) = if !ready {
+            // a new block did not become ready, which means we couldn't have gotten a new cid.
+            (None, Vec::new())
+        } else {
+            // a new leaf must be output, as well as possibly a new link block
+            let leaf = Some(self.flush_buffered_leaf().unwrap());
+            let links = self.flush_buffered_links(NonZeroUsize::new(174).unwrap(), false);
+
+            (leaf, links)
+        };
+
+        Ok((leaf.into_iter().chain(links.into_iter()), written))
```

Comment:
The initial idea for deserialization was to have a 

`fn push<'a>(&'a mut self, input: &[u8]) -> (impl Iterator<Item = (Cid, &'a [u8]), usize)`

or:

`fn push<'a>(&mut self, input: &[u8], cache: &'a mut Option<Cache>) -> (impl Iterator<Item = (Cid, &'a [u8]), usize)`.

At the moment I think we require `Box<[u8]>` for put_block so this reuse would be quite moot. A lifetime-less solution would be to use `bytes::Bytes` which would be reusable with luck.

Another way to speed up here could be with `input.len() >= self.size_hint()` to produce the block without first copying it locally. Not sure if it would help non-size chunkers (which haven't been implemented)..

Reply:
The last idea here was explored above: https://github.com/rs-ipfs/rust-ipfs/pull/220/files#r449066243